When AI rolls "Unique AI" station trait, upload modules won't be available until research, and research requires more points.

tralezab opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

* Unique AI now removes the upload modules from the station
* Unique AI now bumps up the cost of the research that unlocks printing law modules

## Why It's Good For The Game

This was brought up as part of the community meeting, and essentially, it's really easy for admins to miss the fact that when this station trait rollsm the captain almost always switches back to asimov instantly. It was confirmed this was rule breaking, and yet it still happens every single time. I like the idea of being able to naturally earn a way back to asimov for people who are interested, and otherwise we get to actually explore some interesting AI lawsets that aren't insanely station sided, like Hulkamania, Reporter, or even Tyrant... oooh, how exciting...

(This is my second attempt at the issue, the first being really disliked by potato for being too overhanded, hard locking mechanics for reasons that feel too OOC-y. This new attempt is a compromise that I hope everyone can be happy enough with.)

### Some extra stuff to make the trait better as well:

* The game now has a "round_default_lawset" which anything with laws pulls from as its default.
* AI gets a printout of their special laws on rounds with this station trait
* Reset is still in, and there will be modules in the core but they are for the lawset picked. So on normal rounds, they're asimov, but on Unique AI rounds, they are whatever the AI is.


round_default_lawset, ai is warned roundstart, only some modules delete and the asimov module is replaced by the "round default laws" type that assumes the module that is chosen for the round

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl:
balance: Unique AI trait now gets rid of the station's DIFFERENT lawset modules, and increases the research cost to unlocking them.
qol: AI is given a printout of their new laws
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

tralezab committed:
tyrannical code, tyrannical ai

tralezab committed:
station level, more informative report

LemonInTheDark commented:
For the record:
I think the key issue here is basically any ai law change immediately resulting in a reset. I'm just not sure how to cleanly solve that without letting all ai lawsets kill people

tralezab commented:
I have no idea either but I am open to suggestions

necromanceranne commented:
> For the record: I think the key issue here is basically any ai law change immediately resulting in a reset. I'm just not sure how to cleanly solve that without letting all ai lawsets kill people Basically people will gravitate to asimov, and unless there's another human acting to keep asimov off, or the ai can fight back, they'll win.

Remove Asimov as a core AI lawset NOW

Ryusenshu commented:
Would be neat if its possible to change that asimov board in upload depending on the unique ai chosen, so it cant be reset to that

Bumping up Research Cost for AI Tech is just meh though, its already not a priority and trying to fix a compromised ai will just be harder if it just locks down upload

Ryusenshu commented:
> > For the record: I think the key issue here is basically any ai law change immediately resulting in a reset. I'm just not sure how to cleanly solve that without letting all ai lawsets kill people Basically people will gravitate to asimov, and unless there's another human acting to keep asimov off, or the ai can fight back, they'll win.
> 
> Remove Asimov as a core AI lawset NOW

I like the one they have on Shiptest
Its basically just "help people"

necromanceranne commented:
I suppose one solution is 'no asimov, no purge' during unique AI lawset, and making purge an expensive cargo order so that doing it is activately adversarial towards cargo if done early.

Creates a bit of conflict.

tralezab commented:
> > For the record: I think the key issue here is basically any ai law change immediately resulting in a reset. I'm just not sure how to cleanly solve that without letting all ai lawsets kill people Basically people will gravitate to asimov, and unless there's another human acting to keep asimov off, or the ai can fight back, they'll win.
> 
> Remove Asimov as a core AI lawset NOW

Even if you did, there is no chance in sam hell the captain is going to upload anything other than at the least a very heavily station sided lawset every time

tralezab commented:
> I like the one they have on Shiptest Its basically just "help people"

see told you

Ryusenshu commented:
in regards to asimov as base law
can be more about conflicting orders then, but it has nothing to do with the pr right now

Mothblocks commented on code/modules/research/techweb/all_nodes.dm:
```diff
@@ -912,6 +912,12 @@
 	)
 	research_costs = list(TECHWEB_POINT_TYPE_GENERIC = 2500)
 
+//Yes, nodes have their own initialize (thanks kev!) but this should trigger before initialize just fine.
```
Explain why in your comment, not just acknowledging it

zxaber commented:
The reset board should stay, as a tool for removing hacked laws.

The reset board should also be changed to return the AI to the experimental law set if the station trait rolls. Possibly all new AI should also have this law. (Either or both of these may already be true, I haven't tested)

Ryusenshu commented:
> The reset board should stay, as a tool for removing hacked laws.
> 
> The reset board should also be changed to return the AI to the experimental law set if the station trait rolls. Possibly all new AI should also have this lawset. (Either or both of these may already be true, I haven't tested)

I think reset board already does it, but its very ineffective at doing that
Purge and re-apply is way better, since it is actually a clean swipe

Also if unique Law rolls, every unsynced borg or new AI gets a different lawset (its some funny chaos)

CameronWoof commented:
Can someone expand on why fixing the AI laws when it's discovered they're incorrect is against the rules? I can't find anything on discord about it and I also can't imagine what rule is being broken by curating the AI's laws.

BlueMemesauce commented:
> Can someone expand on why fixing the AI laws when it's discovered they're incorrect is against the rules? I can't find anything on discord about it and I also can't imagine what rule is being broken by curating the AI's laws.

The flavor text states "For experimental purposes, this station AI might show divergence from default lawset. Do not meddle with this experiment." So you aren't allowed to meddle with the experiment by changing the laws.

tralezab commented:
> Can someone expand on why fixing the AI laws when it's discovered they're incorrect is against the rules? I can't find anything on discord about it and I also can't imagine what rule is being broken by curating the AI's laws.

For the same reason you can't clear out armory roundstart, or make atmos sabotage proof. It falls under rule 2.

wesoda25 commented:
Only being able to order AI lawsets is actually a really good idea. It's always made me frown a bit how easy it is to covertly change laws.

tralezab committed:
ux improvement package

round_default_lawset, ai is warned roundstart, only some modules delete and the asimov module is replaced by the "round default laws" type that assumes the module that is chosen for the round

tralezab committed:
thank you unit testing, but if you error again...

moocowswag commented:
> The flavor text states "For experimental purposes, this station AI might show divergence from default lawset. Do not meddle with this experiment." So you aren't allowed to meddle with the experiment by changing the laws.

Usually people use these flavor texts as a notification for something wrong that they want to correct.

Like when there is a rare specimen to cultivate in xenobio (a xeno egg) captains will immediately tide xenobio and destroy the egg.

tralezab commented:
>create more issues yet to be foreseen.

Look I'm willing to implement things to work around these issues but feedback like this doesn't give me much help. Speaking of which, I've implemented many improvements, now added to OP

bluetshirtguy commented:
Does this mean Ai's are chained to the mother drone lawset?

zxaber commented:
IIRC the drone lawset is blacklisted from the station trait picking it.

Mothblocks commented:
It's config, but we set it like that yeah

Maurukas reviewed (approved):
Mapping is a straight up replacement and looks fine.

I agree that if the station rolls a unique ruleset, then resetting the AI should restore that ruleset, not set it to Asimov. Otherwise you can just wipe out a unique feature of a round.

Mothblocks commented:
@tralezab Do you wanna do that? If not I can be fine with it being out of scope and just merge as is

tralezab commented:
> 

This already does do that riiiight...

tralezab commented on code/game/objects/items/AI_modules.dm:
```diff
@@ -359,6 +370,20 @@ AI MODULES
 
 /******************** Asimov ********************/
 
+/obj/item/ai_module/core/round_default_laws
+	name = "Default Laws" //gets renamed
+
+/obj/item/ai_module/core/round_default_laws/Initialize(mapload)
+	. = ..()
+	if(!GLOB.round_default_lawset)
+		GLOB.round_default_lawset = setup_round_default_laws()
+	var/datum/ai_laws/default_laws = new GLOB.round_default_lawset()
+	name = "'[default_laws.name]' Core AI Module"
+	laws = default_laws.inherent
+
+/obj/item/ai_module/core/round_default_laws/handle_unique_ai()
+	return
```
here! These replaced asimov

tralezab commented:
conflict

LemonInTheDark commented:
@Mothblocks you happy with this?

Mothblocks commented:
ya

LemonInTheDark commented:
ARMMMMMMMMMMMMM

tralezab commented:
> ARMMMMMMMMMMMMM

![image](https://user-images.githubusercontent.com/40974010/170637085-9209bb53-2602-4849-8fb5-c3eb1e5f9a1d.png)

tralezab committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into a-less-hamfisted-(even-if-it-looks-like-it)-solution

tralezab committed:
this pr be ready

Watermelon914 commented on code/datums/ai_laws/ai_laws.dm:
```diff
@@ -1,5 +1,67 @@
+	if(HAS_TRAIT(SSstation, STATION_TRAIT_UNIQUE_AI))
+		return pick_weighted_lawset()
+
+	switch(CONFIG_GET(number/default_laws))
+		if(0) //plain ol' asimov
+			return /datum/ai_laws/default/asimov
+		if(1) //custom configged laws
+			return /datum/ai_laws/custom
+		if(2) //random laws
+			var/list/randlaws = list()
+			for(var/lpath in subtypesof(/datum/ai_laws))
+				var/datum/ai_laws/L = lpath
+				if(initial(L.id) in law_ids)
+					randlaws += lpath
+			var/datum/ai_laws/lawtype
+			if(randlaws.len)
+				lawtype = pick(randlaws)
+			else
+				lawtype = pick(subtypesof(/datum/ai_laws/default))
+
+			return lawtype
+
+		if(3) //random WEIGHTED laws :)
+			return pick_weighted_lawset()
```
these should have defines

Watermelon914 commented on code/datums/ai_laws/ai_laws.dm:
```diff
@@ -57,56 +111,10 @@
-
-	var/datum/ai_laws/templaws = new lawtype()
-	inherent = templaws.inherent
+	if(!GLOB.round_default_lawset)
+		GLOB.round_default_lawset = setup_round_default_laws()
```
consider initializing this when the global variable is set up

tralezab replied:
Due to order of init, I cannot sadly.

Watermelon914 commented on code/datums/station_traits/neutral_traits.dm:
```diff
@@ -21,9 +21,15 @@
 
+/datum/station_trait/unique_ai/on_round_start()
+	. = ..()
+	for(var/mob/living/silicon/ai/ai in GLOB.ai_list)
+		ai.show_laws()
```
```suggestion
	for(var/mob/living/silicon/ai/ai as anything in GLOB.ai_list)
		ai.show_laws()
```

Watermelon914 commented on code/game/objects/items/AI_modules.dm:
```diff
@@ -357,8 +365,36 @@ AI MODULES
 		law_datum.clear_zeroth_law(0)
 	..()
 
+/obj/item/ai_module/core/full/handle_unique_ai()
+	var/datum/ai_laws/unique_ai_laws =  GLOB.round_default_lawset
+	if(law_id == initial(unique_ai_laws.id))
+		return
+	return QDEL_MODULE
```
consider just returning TRUE/FALSE, no need for a new define here

tralezab replied:
I think it works nicely. I didn't want to copy INIT HINT QDEL ones but I think true doesn't really describe what it will mean. Mothblocks will have a very long rant about it if you ask

Watermelon914 replied:
If you change the name of the function, it will be.
Example:
should_qdel_module

Watermelon914 commented on code/game/objects/items/AI_modules.dm:
```diff
@@ -357,8 +365,36 @@ AI MODULES
+
+/obj/item/ai_module/core/round_default_laws/Initialize(mapload)
+	. = ..()
+	if(!GLOB.round_default_lawset)
+		GLOB.round_default_lawset = setup_round_default_laws()
```
see above

tralezab replied:
Due to order of init, I cannot sadly.

Watermelon914 commented on code/game/objects/items/AI_modules.dm:
```diff
@@ -357,8 +365,36 @@ AI MODULES
+
 /******************** Asimov ********************/
 
+/obj/item/ai_module/core/round_default_laws
+	name = "Default Laws" //gets renamed
```
maybe make this a spawner instead so that it gets deleted either way, that way it isn't confusing.

tralezab replied:
It needs to have something exist to be the round default if it needs to fallback, aka a /datum/lawset that doesn't have a module object with a matching id.

Watermelon914 replied:
you can then spawn the fallback instead

tralezab replied:
Alright if you're sure

tralezab committed:
review!

tralezab committed:
spawner

tralezab committed:
SHOULD

tralezab committed:
get round default

Mothblocks commented on code/datums/ai_laws/ai_laws.dm:
```diff
@@ -2,7 +2,14 @@
 ///Requires config, which isn't loaded at the time global vars init, so setup_round_default_laws is called by the first request to use this global.
-GLOBAL_VAR(round_default_lawset)
+///Do NOT get this directly! It's unsafe! Instead, use `get_round_default_lawset`!
+GLOBAL_VAR(_round_default_lawset)
```
Why not just use a static var inside `get_round_default_lawset`?

tralezab replied:
Shore!

tralezab committed:
ogey
